Separate Direct and Indirect streaming paths, implement indirect path for CRI

timstclair opened the pull request:
This PR refactors the `pkg/kubelet/container.Runtime` interface to remove the `ExecInContainer`, `PortForward` and `AttachContainer` methods. Instead, those methods are part of the `DirectStreamingRuntime` interface which all "legacy" runtimes implement. I also added an `IndirectStreamingRuntime` which handles the redirect path and is implemented by CRI runtimes. To control the size of this PR, I did not fully setup the indirect streaming path for the dockershim, so I left legacy path behind.

Most of this PR is moving & renaming associated with the refactoring. To understand the functional changes, I suggest tracing the code from `getExec` in `pkg/kubelet/server/server.go`, which calls `GetExec` in `pkg/kubelet/kubelet_pods.go` to determine whether to follow the direct or indirect path.

For https://github.com/kubernetes/kubernetes/issues/29579

/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36020)
<!-- Reviewable:end -->

yujuhong commented:
Reviewed 11 of 20 files at r1.
Review status: 11 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kubelet_pods.go, line 1134 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXIRnfGLX3iqL0GqhT:-KVXIRnfGLX3iqL0GqhU:bx85u23) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/kubelet_pods.go#L1134)):*
> ```Go
> 		return nil, err
> 	}
> 	podUID = kl.podManager.TranslatePodUID(podUID)
> ```

I'm fine with moving the function call here, but we should be careful and make sure that all outward facing methods call this function. Ideally we should just use a wrapper for all host method to ensure translation is done. This is out of the scope for the current PR though.

---

*[pkg/kubelet/container/runtime.go, line 124 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXBv4VJxclhSW4JlHe:-KVXBv4Wv7gnlU0BLbIm:blzbj4) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L124)):*
> ```Go
> }
> 
> type DirectStreamingRuntime interface {
> ```

Add comment for the interface. 

---

*[pkg/kubelet/container/runtime.go, line 135 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXByp9Wu2XiMRaj8PS:-KVXByp9Wu2XiMRaj8PT:blzbj4) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L135)):*
> ```Go
> }
> 
> type IndirectStreamingRuntime interface {
> ```

Add comment for the interface. 

---

*[pkg/kubelet/container/runtime.go, line 138 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXFvxJQvOe7xlvFn6A:-KVXFvxJQvOe7xlvFn6B:b-qufy4x) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L138)):*
> ```Go
> 	GetExec(id ContainerID, cmd []string, stdin, stdout, stderr, tty bool) (*url.URL, error)
> 	GetAttach(id ContainerID, stdin, stdout, stderr bool) (*url.URL, error)
> 	GetPortForward(podFullName string, podUID types.UID) (*url.URL, error)
> ```

Can we not use `podFullName`? It's not very well-defined. "UID" or "name+namespace" is preferred. e

---

*[pkg/kubelet/dockertools/docker_manager.go, line 267 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXI3Qx28h6MZRbH2d6:-KVXI3Qx28h6MZRbH2d7:b-ldp597) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/dockertools/docker_manager.go#L267)):*
> ```Go
> 		imageStatsProvider:     newImageStatsProvider(client),
> 		seccompProfileRoot:     seccompProfileRoot,
> 	}
> ```

Need to ensure that docker manager implements `DirectStreamingRuntime` now that it's not part of the `Runtime` interface. 
Add `var _ kubecontainer.DirectStreamingRuntime = &DockerManager{}` to the beginning of the file.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 116 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXNS9EaK6ij7hM2UlB:-KVXNS9FMHBFk9SjwapQ:b8by7g7) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L116)):*
> ```Go
> 	kubecontainer.Runtime
> 	kubecontainer.IndirectStreamingRuntime
> 	kubecontainer.ContainerCommandRunner
> ```

include `kubecontainer.DirectStreamingRuntime` since we still keep the old methods.

---

*[pkg/kubelet/rkt/rkt.go, line 278 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXHT0FkDlouCezvUa4:-KVXHT0FkDlouCezvUa5:b-xuhv0m) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/rkt/rkt.go#L278)):*
> ```Go
> 		return nil, fmt.Errorf("rkt: cannot get config from rkt api service: %v", err)
> 	}
> ```

Need to ensure that rkt implements `DirectStreamingRuntime` now that it's not part of the `Runtime` interface. 
Add `var _ kubecontainer.DirectStreamingRuntime = &Runtime{}` to the beginning of the file.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Made the first pass. Most changes seem reasonable/straightforward. The cleanup of `RunInContainer` is nice. Will take a closer look later.  Look forward to the next PR where legacy methods get deleted.

timstclair commented:
Thanks, addressed all comments.

---

Review status: 11 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kubelet_pods.go, line 1134 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXIRnfGLX3iqL0GqhT:-KVXebWJdg1jUI5xj3q9:b-h6ak1o) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/kubelet_pods.go#L1134)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> I'm fine with moving the function call here, but we should be careful and make sure that all outward facing methods call this function. Ideally we should just use a wrapper for all host method to ensure translation is done. This is out of the scope for the current PR though.
</details>
I double checked that all the places I removed `TranslatePodUID(...)` from call findContainer, and don't use the PodUID for anything else. TranslatePodUID should be idempotent, so if it's called by mistake before calling findContainer it won't be an issue. Agree about leveraging the type system to keep track of this transformation, filed https://github.com/kubernetes/kubernetes/issues/36031

---

*[pkg/kubelet/container/runtime.go, line 124 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXBv4VJxclhSW4JlHe:-KVXgar0hBksQR2NsoT7:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L124)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Add comment for the interface. 
</details>
Done.

---

*[pkg/kubelet/container/runtime.go, line 135 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXByp9Wu2XiMRaj8PS:-KVXgb70-lplPpgISLuH:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L135)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Add comment for the interface. 
</details>
Done.

---

*[pkg/kubelet/container/runtime.go, line 138 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXFvxJQvOe7xlvFn6A:-KVXlNZAnmi4y78Pnxef:b-ho5wpv) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/container/runtime.go#L138)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Can we not use `podFullName`? It's not very well-defined. "UID" or "name+namespace" is preferred. e
</details>
Agreed. Done.

---

*[pkg/kubelet/dockertools/docker_manager.go, line 267 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXI3Qx28h6MZRbH2d6:-KVXl_4dLALA4_Ov_Opi:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/dockertools/docker_manager.go#L267)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Need to ensure that docker manager implements `DirectStreamingRuntime` now that it's not part of the `Runtime` interface. 
> Add `var _ kubecontainer.DirectStreamingRuntime = &DockerManager{}` to the beginning of the file.
</details>
Done.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 116 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXNS9EaK6ij7hM2UlB:-KVXlpg6d0CeVRI7yjFG:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L116)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> include `kubecontainer.DirectStreamingRuntime` since we still keep the old methods.
</details>
Done.

---

*[pkg/kubelet/rkt/rkt.go, line 278 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-KVXHT0FkDlouCezvUa4:-KVXlzBMHBlJ36rXk8gk:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c53a548fb9463ecc9e204073997443f419fa47a2/pkg/kubelet/rkt/rkt.go#L278)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Need to ensure that rkt implements `DirectStreamingRuntime` now that it's not part of the `Runtime` interface. 
> Add `var _ kubecontainer.DirectStreamingRuntime = &Runtime{}` to the beginning of the file.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36020#-:-KVXm9wslBn25N8ntFBB:bhposek)*
<!-- Sent from Reviewable.io -->

feiskyer commented on pkg/kubelet/container/runtime.go:
```diff
@@ -124,6 +121,28 @@ type Runtime interface {
+type IndirectStreamingRuntime interface {
+	GetExec(id ContainerID, cmd []string, stdin, stdout, stderr, tty bool) (*url.URL, error)
+	GetAttach(id ContainerID, stdin, stdout, stderr bool) (*url.URL, error)
+	GetPortForward(podName, podNamespace string, podUID types.UID) (*url.URL, error)
```
Where is `port`? Is it handled somewhere else?

timstclair replied:
Unfortunately the ports aren't included in the initial request, so without some changes to the client API we don't have that information here. I left a TODO [here](https://github.com/kubernetes/kubernetes/pull/36020/files/b97fc5c794374cb6c83a04b4fd7d89ab2a3762fe#diff-0170c1534446e377b025a1c54ab478e2R234) to deal with it later.

feiskyer commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -181,6 +181,7 @@ type Runtime struct {
 }
 
 var _ kubecontainer.Runtime = &Runtime{}
+var _ kubecontainer.DirectStreamingRuntime = &Runtime{}
```
Have a question: is there any initialization problem of assigning `kubecontainer.DirectStreamingRuntime` in both dockertools and rkt package?

timstclair replied:
This is just asserting that `&Runtime{}` implements the `kubecontainer.DirectStreamingRuntime` interface. Nothing is being assigned.

yujuhong commented:
LGTM

timstclair commented:
Squashed & rebased.

timstclair commented:
Added comment to fix `hack/verify-golint.sh`. Reapplying LGTM.

euank commented:
@k8s-bot verify test this issue: #36110

euank commented:
@timstclair 
```
W1102 16:39:56.277] Errors from golint:
W1102 16:39:56.277] pkg/kubelet/util/format/pod.go:36:1: comment on exported function PodDesc should be of the form "PodDesc ..."
W1102 16:39:56.277] 
W1102 16:39:56.278] Please fix the above errors. You can test via "golint" and commit the result.
W1102 16:39:56.278] 
```

Belated LGTM btw, I gave it a go and got it the exec part of it at least to work e2e for me.

tallclair committed:
Separate Direct and Indirect streaming paths, implement indirect path with CRI

timstclair commented:
Thanks, fixed. Reapplying LGTM (trivial comment change).

timstclair commented:
@k8s-bot verify test this

yujuhong commented:
@timstclair 
```
W1102 19:53:41.694] pkg/kubelet/kubelet.go:622: k8s.io/kubernetes/pkg/kubelet/container.ContainerCommandRunnerWrapper composite literal uses unkeyed fields
W1102 19:53:41.694] pkg/kubelet/kubelet.go:654: k8s.io/kubernetes/pkg/kubelet/container.ContainerCommandRunnerWrapper composite literal uses unkeyed fields
W1102 19:53:41.737] exit status 1
W1102 19:53:43.070] pkg/kubelet/dockertools/docker_manager.go:269: k8s.io/kubernetes/pkg/kubelet/container.ContainerCommandRunnerWrapper composite literal uses unkeyed fields
W1102 19:53:43.092] exit status 1
W1102 19:53:44.950] pkg/kubelet/rkt/rkt.go:280: k8s.io/kubernetes/pkg/kubelet/container.ContainerCommandRunnerWrapper composite literal uses unkeyed fields
```

timstclair commented:
I can't tell what's failing in verify. I just sync'd my client to see if it was resolved.

timstclair commented:
I still can't tell what's failing, but I tried cleaning up the unkeyed field warnings.

timstclair commented:
Hmm, that seemed to fix the verify error. Squashed the commit. PTAL.

yujuhong commented:
@k8s-bot gci gce e2e test this

yujuhong commented:
@timstclair I think this PR broke the CRI tests.
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubelet-cri-gce-e2e-ci/842
```
Liveness probe for "liveness-exec_e2e-tests-container-probe-zs8nq(a47cd450-a2d2-11e6-bc63-42010a800007):liveness" errored: rpc error: code = 2 desc = not implemented
```
I should've triggered the CRI tests with `@k8s-bot cri test this`. My bad. Will look into this after the meeting.

feiskyer commented:
> Liveness probe for "liveness-exec_e2e-tests-container-probe-zs8nq(a47cd450-a2d2-11e6-bc63-42010a800007):liveness" errored: rpc error: code = 2 desc = not implemented

@timstclair Will this issue be fixed by #36253?

Random-Liu commented:
> @timstclair Will this issue be fixed by #36253?

Just saw your comment. Yeah, it seems so. But I've already sent a quick fix. Haha
